x/reward/keeper/query_server.go

```diff
@@ -0,0 +1,203 @@
+		var claimPeriodRewardDist types.ClaimPeriodRewardDistribution
+		vErr := claimPeriodRewardDist.Unmarshal(value)
+		if accumulate {
+			if vErr == nil {
+				response.ClaimPeriodRewardDistributions = append(response.ClaimPeriodRewardDistributions, claimPeriodRewardDist)
+			} else {
+				return false, vErr
+			}
+		}
+		return true, nil
+	})
+	if err != nil {
+		return &response, status.Error(codes.Unavailable, err.Error())
+	}
+	response.Pagination = pageRes
+	return &response, nil
+}
+
+// ClaimPeriodRewardDistributionsByID returns a ClaimPeriodRewardDistribution by rewardID and claimPeriodID
+func (k Keeper) ClaimPeriodRewardDistributionsByID(ctx context.Context, req *types.QueryClaimPeriodRewardDistributionsByIDRequest) (*types.QueryClaimPeriodRewardDistributionsByIDResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	response := types.QueryClaimPeriodRewardDistributionsByIDResponse{}
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	ClaimPeriodReward, err := k.GetClaimPeriodRewardDistribution(sdkCtx, req.GetClaimPeriodId(), req.GetRewardId())
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, fmt.Sprintf("unable to query claim period reward distributions by ID: %v", err))
+	}
+
+	if ClaimPeriodReward.Validate() == nil {
+		response.ClaimPeriodRewardDistribution = &ClaimPeriodReward
+	}
+
+	return &response, nil
+}
+
+func (k Keeper) RewardDistributionsByAddress(ctx context.Context, request *types.QueryRewardDistributionsByAddressRequest) (*types.QueryRewardDistributionsByAddressResponse, error) {
+	if request == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+	address, err := sdk.AccAddressFromBech32(request.Address)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, err.Error())
+	}
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	var states []types.RewardAccountState
+	getAllRewardAccountStore := prefix.NewStore(sdk.UnwrapSDKContext(ctx).KVStore(k.storeKey), types.GetAllRewardAccountByAddressPartialKey(address))
+
+	pageRes, err := query.FilteredPaginate(getAllRewardAccountStore, request.Pagination, func(key []byte, value []byte, accumulate bool) (bool, error) {
+		lookupVal, errFromParsingKey := types.ParseFilterLookUpKey(key, address)
+		if errFromParsingKey != nil {
+			return false, err
+		}
+		if accumulate {
+			result, errFromGetRewardAccount := k.GetRewardAccountState(sdkCtx, lookupVal.RewardID, lookupVal.ClaimID, lookupVal.Addr.String())
+			// think ignoring the error maybe ok here since it's just another lookup
+			if errFromGetRewardAccount != nil {
+				return false, errFromGetRewardAccount
+			}
+			if result.GetSharesEarned() == 0 || (request.ClaimStatus != result.ClaimStatus && request.ClaimStatus != types.RewardAccountState_CLAIM_STATUS_UNSPECIFIED) {
+				return false, nil
+			}
+
+			states = append(states, result)
+		}
```

Comment:
Did this get changed recently? I could have sworn it used to only have the `states = append(states, result)` inside the `if accumulate {` block.

If there's an entry where `errFromGetRewardAccount` is nil or the other `if` stuff is true, then the counting will be off when using an offset for pagination. This will specially be noticeable if filtering the results by claim status. I know it sucks that all the extra work needs to be done when not accumulating, but in this case, there's no way around it; it all needs to be done before we know whether or not it should count as an entry.

Reply:
>  I could have sworn it used to only have the states = append(states, result) inside the if accumulate { block.

still seeing it happen within the loop, unless i am misunderstanding your comment.

Reply:
I moved the accumulate method back.  I understand it better now.  Thanks.

Reply:
Good catch!  Thanks!